### PR TITLE
[feature] Migrate from cron to celery_beat

### PR DESCRIPTION
**Pull request by cslzchen**
Move shellscript-based crontab tasks to celery periodic tasks using celery beat. Here is the [progress report](https://docs.google.com/document/d/1H3PDHimqVlmvS4bFHhjOuBJCzpn8dAkvbuF2dV0T67A/edit?usp=sharing).

### CODE REVIEW

File: scripts/osfstorage/files_audit.py [hunk 1]
```diff
@@ -12,14 +12,20 @@
...
 import os
 import math
+import thread
 import logging
 
 import pyrax
 import progressbar
+
 from modularodm import Q
+
```
Comment (icereval) [inline scripts/osfstorage/files_audit.py]: try to keep non-std lib, exsternal modules grouped together when possible

File: scripts/triggered_mails.py [hunk 2]
```diff
@@ -1,20 +1,24 @@
...
-from framework.auth import User
 
 from scripts.utils import add_file_logger
 
 logger = logging.getLogger(__name__)
 logging.basicConfig(level=logging.INFO)
 
+
 def main(dry_run=True):
+    print('trigger_mail')
```
Comment (icereval) [inline scripts/triggered_mails.py]: probably used for testing

File: scripts/refresh_box_tokens.py [hunk 3]
```diff
@@ -42,15 +46,15 @@ def main(delta, dry_run):
                 logger.error(ex.message)
 
 
-if __name__ == '__main__':
+@celery_app.task(name='scripts.refresh_box_tokens')
+def run_main(days, dry_run):
```
Comment (icereval) [inline scripts/refresh_box_tokens.py]: `dry_run=True`

Comment (icereval) [inline scripts/refresh_box_tokens.py]: `days=None`

File: website/settings/defaults.py [hunk 4]
```diff
@@ -317,8 +333,59 @@ def parent_dir(path):
             'schedule': crontab(minute=0, hour=0),
             'args': ('email_digest',),
         },
+        'refresh_box': {
+            'task': 'scripts.refresh_box_tokens',
+            'schedule': crontab(minute=0, hour= 2),  # Daily 2:00 a.m
+            'args': (DAYS, DRY_RUN,),
```
Comment (icereval) [inline website/settings/defaults.py]: once the above default values are set, these args can be simply `dry_run=False`.

Comment (icereval) [inline website/settings/defaults.py]: This can be done for the bellow function calls as well.

File: scripts/osfstorage/files_audit.py [hunk 5]
```diff
@@ -127,11 +134,19 @@ def main(nworkers, worker_id, dry_run):
     progress_bar.finish()
 
 
-if __name__ == '__main__':
-    import sys
-    nworkers = int(sys.argv[1])
-    worker_id = int(sys.argv[2])
-    dry_run = 'dry' in sys.argv
+@celery_app.task(name='scripts.osfstorage.files_audit')
+def run_mains(num_of_workers, dry_run):
```
Comment (icereval) [inline scripts/osfstorage/files_audit.py]: default to 4

File: scripts/osfstorage/files_audit.py [hunk 6]
```diff
@@ -127,11 +134,19 @@ def main(nworkers, worker_id, dry_run):
...
+    try:
+        thread.start_new_thread(run_main, (num_of_workers, 0, dry_run,))
```
Comment (icereval) [inline scripts/osfstorage/files_audit.py]: Seems like this might need to be a `for` loop over the number of worker's specified.

Comment (icereval) [inline scripts/osfstorage/files_audit.py]: It appears this task might be better spawning additional Celery tasks, so you won't need to manage the threads.

File: scripts/osfstorage/usage_audit.py [hunk 7]
```diff
@@ -95,9 +99,10 @@ def main(send_email=False):
         logger.info('No offending projects or users found')
 
 
-if __name__ == '__main__':
+@celery_app.task(name='scripts.osfstorage.usage_audit')
+def run_main(*arg):
```
Comment (icereval) [inline scripts/osfstorage/usage_audit.py]: `args` can be converted into specific method inputs

**Comment by icereval** [comment]
First pass complete, :palm_tree:

**Comment by cslzchen** [comment]
please ignore this update, I just realize that pushes to the repo with open pull request will be automatically updated

**Comment by cslzchen** [comment]
Add analytics to celery beat and fix several None/Empty bug. Ready for review.

**Comment by cslzchen** [comment]
Add uploading analytics results to OSFStorage.
Add fix to handle server 503 error
Support the latest waterbutler api update (yesterday)
Ready for review

File: requirements/release.txt [hunk 8]
```diff
@@ -3,7 +3,7 @@
 # Requirements to be installed on server deployments
 
 # scripts osfstorage
-pyrax==1.9.5
+pyrax==1.9.6
```
Comment (cwisecarver) [inline requirements/release.txt]: There's a 1.9.7 released now that "fixes a pypi packaging issue"

File: scripts/analytics/upload.py [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+            if test_existence:
+                return False
+            else:
+                create_dir(a_dir, None, parent_path, parent_id)
+                dirs = database['storedfilenode'].find({'name': a_dir, 'parent': parent_id})
+                if dirs.count() == 1:
+                    parent_id = dirs[0]['_id']
+                    parent_path.append(parent_id)
+                else:
+                    print('I/O Error: cannot find newly created directory')
```
Comment (chrisseto) [inline scripts/analytics/upload.py]: Should this just be an exception?
`raise Exception('Cannot find newly created directory')`

File: scripts/analytics/upload.py [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+            parent_id = dirs[0]['_id']
+            parent_path.append(parent_id)
+            continue
+
+    try:
+        file_path = '/' + '/'.join(website_settings.ANALYTICS_PATH.split('/')[1:-1]) + path + '/' + name
+        print('file_path: {}'.format(file_path))
+        file_stream = open(file_path, 'r')
+    except IOError as e:
+        print('I/O error({0}): {1}'.format(e.errno, e.strerror))
```
Comment (chrisseto) [inline scripts/analytics/upload.py]: Any reason to not just raise this exception?

File: scripts/analytics/upload.py (hunk 9 above)
Comment (cslzchen) [inline scripts/analytics/upload.py]: @chrisseto You are right, thanks. I had some tested code here. Forgot to change back to "raise Exception". Same for the code below.

File: requirements/release.txt (hunk 8 above)
Comment (cslzchen) [inline requirements/release.txt]: @cwisecarver I will update this in the next pull request.

**Comment by chrisseto** [comment]
@cslzchen Looks good to me.

We might want to document this somewhere as I don't see it documented on celery's docs; tasks can be manually invoked with `celery call`

```bash
~/D/C/osf ❯❯❯ celery call --help                                                                
Usage: celery call [options] <task_name>

Call a task by name.

Examples::

    celery call tasks.add --args='[2, 2]'
    celery call tasks.add --args='[2, 2]' --countdown=10

Options:
  -A APP, --app=APP     app instance to use (e.g. module.attr_name)
  -b BROKER, --broker=BROKER
                        url to broker.  default is 'amqp://guest@localhost//'
  --loader=LOADER       name of custom loader class to use.
  --config=CONFIG       Name of the configuration module
  --workdir=WORKING_DIRECTORY
                        Optional directory to change to after detaching.
  -C, --no-color
  -q, --quiet
  -a ARGS, --args=ARGS  positional arguments (json).
  -k KWARGS, --kwargs=KWARGS
                        keyword arguments (json).
  --eta=ETA             scheduled time (ISO-8601).
  --countdown=COUNTDOWN
                        eta in seconds from now (float/int).
  --expires=EXPIRES     expiry time (ISO-8601/float/int).
  --serializer=SERIALIZER
                        defaults to json.
  --queue=QUEUE         custom queue name.
  --exchange=EXCHANGE   custom exchange name.
  --routing-key=ROUTING_KEY
                        custom routing key.
  --version             show program's version number and exit
  -h, --help            show this help message and exit
```

**Pull request closed without merging**